tests/integration/test_concurrent_ttl_merges/test.py

```diff
@@ -0,0 +1,86 @@
+import time
+import pytest
+
+import helpers.client as client
+from helpers.cluster import ClickHouseCluster
+from helpers.test_tools import TSV
+from helpers.test_tools import assert_eq_with_retry
+
+
+cluster = ClickHouseCluster(__file__)
+node1 = cluster.add_instance('node1', main_configs=['configs/fast_background_pool.xml'], with_zookeeper=True)
+node2 = cluster.add_instance('node2', main_configs=['configs/fast_background_pool.xml'], with_zookeeper=True)
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+
+        yield cluster
+
+    finally:
+        cluster.shutdown()
+
+
+def count_ttl_merges_in_queue(node, table):
+    result = node.query("SELECT count() FROM system.replication_queue WHERE merge_type = 'TTL_DELETE' and table = '{}'".format(table))
+    if not result:
+        return 0
+    return int(result.strip())
+
+
+def count_regular_merges_in_queue(node, table):
+    result = node.query("SELECT count() FROM system.replication_queue WHERE merge_type = 'REGULAR' and table = '{}'".format(table))
+    if not result:
+        return 0
+    return int(result.strip())
+
+
+def count_ttl_merges_in_background_pool(node, table):
+    result = node.query("SELECT count() FROM system.merges WHERE merge_type = 'TTL_DELETE' and table = '{}'".format(table))
+    if not result:
+        return 0
+    return int(result.strip())
+
+
+def count_regular_merges_in_background_pool(node, table):
+    result = node.query("SELECT count() FROM system.merges WHERE merge_type = 'REGULAR' and table = '{}'".format(table))
+    if not result:
+        return 0
+    return int(result.strip())
+
+
+def count_running_mutations(node, table):
+    result = node.query("SELECT count() FROM system.merges WHERE table = '{}' and is_mutation=1".format(table))
+    if not result:
+        return 0
+    return int(result.strip())
+
+
+# This test was introduced to check concurrency for TTLs merges and mutations
+# but it revealed a bug when we assign different merges to the same part
+# on the borders of partitions.
+def test_no_ttl_merges_in_busy_pool(started_cluster):
```

Comment:
Actually crash reproduces only in debug build. In other modes it will hang until timeout.